RC/RS: Fully Respect ControllerRef

enisoc opened the pull request:
**What this PR does / why we need it**:

This is part of the completion of the [ControllerRef](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/controller-ref.md) proposal. It brings ReplicaSet and ReplicationController into full compliance with ControllerRef. See the individual commit messages for details.

**Which issue this PR fixes**:

Although RC/RS had partially implemented ControllerRef, they didn't use it to determine which controller to sync, or to update expectations. This could lead to instability or controllers getting stuck.

Ref: https://github.com/kubernetes/kubernetes/issues/24433

**Special notes for your reviewer**:

**Release note**:
```release-note
```
cc @erictune @kubernetes/sig-apps-pr-reviews

Commit:
RC/RS: Use ControllerRef to route watch events.

This is part of the completion of ControllerRef, as described here:

https://github.com/kubernetes/community/blob/master/contributors/design-proposals/controller-ref.md#watches

This also removes the need for the Pod->Controller mapping cache in RC
and RS. This mapping is now persisted in the Pod's ControllerRef
instead.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/41984)
<!-- Reviewable:end -->

kargakis commented on pkg/controller/replicaset/replica_set.go:
```diff
@@ -591,31 +579,22 @@ func (rsc *ReplicaSetController) syncReplicaSet(key string) error {
 		return nil
 	}
 
+	// list all pods to include the pods that don't match the rs`s selector
```
There is already an open PR for this one: https://github.com/kubernetes/kubernetes/pull/41282

krmayankk replied:
@enisoc do you mind leaving out the changes that are already in my PR that kargakis pointed above ?

enisoc replied:
Thanks I'll follow the progress on that one and resolve conflicts after it merges.

kargakis commented on pkg/controller/replicaset/replica_set.go:
```diff
@@ -332,28 +273,48 @@ func (rsc *ReplicaSetController) updatePod(old, cur interface{}) {
+	curControllerRef := controller.GetControllerOf(curPod)
+	oldControllerRef := controller.GetControllerOf(oldPod)
+	controllerRefChanged := !reflect.DeepEqual(curControllerRef, oldControllerRef)
+	if controllerRefChanged {
```
```go
if controllerRefChanged && oldControllerRef != nil {
```
to reduce indentation

enisoc replied:
Done.

kargakis commented:
Do we also need to update lister expansions to respect owner references? I would say yes
https://github.com/kubernetes/kubernetes/blob/17375fc59fff39135af63bd1750bb07c36ef873b/pkg/client/listers/extensions/v1beta1/replicaset_expansion.go#L39

enisoc commented:
@kargakis Regarding `GetPodReplicaSets()`, this is still an operation that we need - list all the RS that match. For example, I use it in this PR to sync matching controllers when an orphan is seen. However, I agree that it might not be what callers want anymore. For some cases they should probably use `GetControllerOf()` instead. I think I'll have to look at each caller on a case-by-case basis.

kargakis commented:
> @kargakis Regarding GetPodReplicaSets(), this is still an operation that we need - list all the RS that match. For example, I use it in this PR to sync matching controllers when an orphan is seen.

But the orphan doesn't have the owner reference right? We can always fallback to listing by labels but I think we can also use owner references in that lister call.

enisoc commented:
Hm I see what you mean. We would only need to separate the two behaviors if there's a use case where you want to get the list of all matching controllers on a non-orphan. I don't see a very large number of callers, so I will take a look at what they're using this for and see what makes sense. Thanks.

enisoc commented:
@kargakis The only uses of `GetPodControllers()` and `GetPodReplicaSets()` I found that need to be changed are the ones in the Disruption controller. I would prefer to do that as a separate PR after all controllers (that Disruption looks at) use ControllerRef, so I can update Disruption holistically for a fully-ControllerRef-ized world.

Commit:
RC/RS: Update Lister documentation for ControllerRef.

The RC/RS Listers still use selectors, because this is the behavior
expected by callers. This clarifies the meaning of the returned list.
Some callers may need to switch to using GetControllerOf() instead,
but that is a separate, case-by-case issue.

kargakis commented:
sgtm

On Thu, Feb 23, 2017 at 7:39 PM, Anthony Yeh <notifications@github.com>
wrote:

> @kargakis <https://github.com/kargakis> The only uses of
> GetPodControllers() and GetPodReplicaSets() I found that need to be
> changed are the ones in the Disruption controller. I would prefer to do
> that as a separate PR after all controllers (that Disruption looks at) use
> ControllerRef, so I can update Disruption holistically for a
> fully-ControllerRef-ized world.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/41984#issuecomment-282080778>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADuFf2IphsolKU0N2qXKmrIhcuVC3o2eks5rfdJOgaJpZM4MKP_t>
> .
>

janetkuo commented on pkg/controller/replicaset/replica_set.go:
```diff
@@ -485,22 +480,15 @@ func (rsc *ReplicaSetController) manageReplicas(filteredPods []*v1.Pod, rs *exte
+					Kind:       getRSKind().Kind,
+					Name:       rs.Name,
+					UID:        rs.UID,
+					Controller: &isController,
```
Will `isController` ever be `false`? If we just want a &true, I'd suggest adding `func BoolPtr` to pkg/util/util.go and just use `BoolPtr(true)`. Or `const isController = true`.

enisoc replied:
This part of my PR is going to disappear anyway because someone else did this in another PR and I'm going to rebase onto that after it merges.

Aside from that, I'm not sure it's worth adding a dependency on another package just to get the Go equivalent of `new bool(true)`. I think it's clear enough that we are creating a variable just above and providing a pointer to that variable.

kow3ns commented:
Reviewed 13 of 16 files at r1, 3 of 3 files at r2, 7 of 7 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41984)*
<!-- Sent from Reviewable.io -->

janetkuo commented on pkg/controller/replicaset/replica_set_test.go:
```diff
@@ -650,16 +653,41 @@ func TestUpdatePods(t *testing.T) {
+				t.Errorf("Expected keys %#v got %v", expected, got)
+			}
+		case <-time.After(wait.ForeverTestTimeout):
+			t.Errorf("Expected update notifications for replica sets within 100ms each")
```
Maybe I missed it. Where does `100ms` come from?

enisoc replied:
Actually I don't know. I copied this from other tests in this file. Maybe the text of the error is outdated. I will make it less specific.

janetkuo commented:
2 comments; LGTM otherwise

kow3ns commented:
/lgtm

kow3ns commented:
/lgtm

kow3ns commented:
Reviewed 4 of 4 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41984)*
<!-- Sent from Reviewable.io -->

Commit:
ControllerRefManager: Don't always filter inactive Pods.

Some controllers, like DaemonSet, want to see all Pods.

Commit:
RC/RS: Don't requeue on error inside sync function.

Returning an error from the sync function already triggers a requeue
in processNextWorkItem().

kow3ns commented:
@k8s-bot cvm gce e2e test this

kow3ns commented:
@k8s-bot gce etcd3 e2e test this

janetkuo commented:
/lgtm

janetkuo commented:
/approve

deads2k commented:
/approve

kargakis commented:
Adding the do-not-merge label until https://github.com/kubernetes/kubernetes/pull/41282 goes in first.

kow3ns commented:
*[pkg/controller/replication/replication_controller.go, line 59 at r6](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41984#-Ke06DMxL_3IHrjBrKfB:-Ke06DMxL_3IHrjBrKfC:b-a5ystt) ([raw file](https://github.com/kubernetes/kubernetes/blob/cf25b8c59882ef05e3423109fbdca64c0fe8495c/pkg/controller/replication/replication_controller.go#L59)):*
> ```Go
> 
> // ControllerKind contains the schema.GroupVersionKind for this controller type.
> var ControllerKind = v1.SchemeGroupVersion.WithKind("ReplicationController")
> ```

Definitely more efficient to use a var, but are you sure you want to export a global mutable var?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41984)*
<!-- Sent from Reviewable.io -->

kow3ns commented:
Reviewed 2 of 4 files at r5, 5 of 5 files at r6.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41984)*
<!-- Sent from Reviewable.io -->

kargakis commented on pkg/controller/replication/replication_controller.go:
```diff
@@ -575,9 +576,6 @@ func (rm *ReplicationManager) syncReplicationController(key string) error {
 	cm := controller.NewPodControllerRefManager(rm.podControl, rc, labels.Set(rc.Spec.Selector).AsSelectorPreValidated(), controllerKind)
 	filteredPods, err = cm.ClaimPods(pods)
 	if err != nil {
-		// Something went wrong with adoption or release.
```
There is one additional enqueue in the error check above this one that is not needed.

kargakis commented on pkg/controller/replication/replication_controller.go:
```diff
@@ -386,31 +386,32 @@ func (rm *ReplicationManager) enqueueControllerAfter(obj interface{}, after time
-		utilruntime.HandleError(err)
-		return false
-	}
 	for {
```
Can you follow the pattern used in the replica set controller and split this out into a separate func? Also invert the returned booleans.

kargakis replied:
Ehm, you just did that. Just the booleans need to be synchronized.

kargakis replied:
Also the way the for loop runs

kargakis commented:
/lgtm

kargakis commented on pkg/controller/replication/replication_controller.go:
```diff
@@ -275,33 +204,49 @@ func (rm *ReplicationManager) updateRC(old, cur interface{}) {
-		utilruntime.HandleError(fmt.Errorf("Couldn't get key for replication controller %#v: %v", rc, err))
-		return
-	}
+	glog.V(4).Infof("Pod %s created: %#v.", pod.Name, pod)
```
We shouldn't log this if the pod belongs to a non-RC controller.

kargakis commented on pkg/controller/replication/replication_controller.go:
```diff
@@ -311,6 +256,7 @@ func (rm *ReplicationManager) updatePod(old, cur interface{}) {
 		return
 	}
 	glog.V(4).Infof("Pod %s updated, objectMeta %+v -> %+v.", curPod.Name, oldPod.ObjectMeta, curPod.ObjectMeta)
```
Basically we shouldn't log these if the pod does not belong to a RC at all. I have been debugging a statefulset failure and it's pretty annoying that this comes up frequently in logs.

caesarxuchao commented on pkg/controller/replicaset/replica_set.go:
```diff
@@ -59,9 +58,8 @@ const (
-	return v1beta1.SchemeGroupVersion.WithKind("ReplicaSet")
-}
+// controllerKind contains the schema.GroupVersionKind for this controller type.
+var controllerKind = v1beta1.SchemeGroupVersion.WithKind("ReplicaSet")
```
It was written as a function to make the returned value "constant". Is there a reason to change it to a var?

enisoc replied:
There is unfortunately a lot of similar-but-not-yet-shareable code between all the controllers, and I wanted to introduce a convention of having `controllerKind` in the package so I can copy code between controllers and it would understand how to ask "what is my kind?".

I figured it serves a similar purpose to the `SchemeGroupVersion` variable itself:

https://github.com/kubernetes/kubernetes/blob/dd22743b54f280f41e68f206449a13ca949aca4e/pkg/apis/extensions/v1beta1/register.go#L29

In that case, not only is the read-only nature merely a convention, but the variable is package public. So I used it as precedent that read-only-by-convention variables are ok. If that's the case, then making it a function seems like unnecessary overhead both cognitively and at runtime.

Commit:
RC/RS: Don't log Pod events unless some RC/RS actually cares.

caesarxuchao commented on pkg/controller/replicaset/replica_set_test.go:
```diff
@@ -674,17 +680,20 @@ func TestUpdatePods(t *testing.T) {
 	pod2.ResourceVersion = "2"
 	manager.updatePod(&pod1, &pod2)
-	expected = sets.NewString(testRSSpec1.Name)
+	expected = sets.NewString(testRSSpec2.Name)
```
Shouldn't rs1 as the previous controller be notified as well?

caesarxuchao replied:
Never mind, I saw the controller ref was set to rs2.

caesarxuchao commented on pkg/controller/replicaset/replica_set_test.go:
```diff
@@ -693,7 +702,52 @@ func TestUpdatePods(t *testing.T) {
+		}
+	}
+
+	// case 2: Remove ControllerRef (orphan). Expect to sync both former owner and
```
typo: s/2/3/

caesarxuchao commented on pkg/controller/replicaset/replica_set_test.go:
```diff
@@ -693,7 +702,52 @@ func TestUpdatePods(t *testing.T) {
+	pod2 = pod1
+	pod2.Labels = labelMap2
+	pod2.ResourceVersion = "2"
+	manager.updatePod(&pod1, &pod2)
```
if the test case is changing label from labelMap2 to labelMap1 (i.e., changing from match controller's selector to not match), is the expected result the same?

calebamiles commented:
@k8s-bot unit test this
@k8s-bot kops aws e2e test this

gmarek commented:
If I'm not mistaken this broke https://k8s-gubernator.appspot.com/builds/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-etcd3/

gmarek commented:
I'm going to revert that and see if it helps.

gmarek commented:
It's possible that #42458 will fix it, so let's try that before reverting.
